### PR TITLE
Remove Python 3.8 and 3.9, add Python 3.11

### LINKED ISSUE #51: Upgrade repo to work with Python 3.11
This repo currently works with Python 3.8-3.10. This is in part checked via GitHub Actions, which runs all unit tests on these versions of Python. As Python 3.8 is no longer supported (as of 07/10/2024) and Python 3.12 is the current stable release, this repo needs to be upgrades asap to work with Python 3.11, at minimum.

### PR DESCRIPTION
**Pull request by mary-cleaton**
## Description

Remove Python 3.8 and 3.9 (as they are unsupported) and add Python 3.11 (as it is the most recent version of Python that works with Spark 3.5.x).

closes #51

## Type of change

- [ ] Bug fix - *non-breaking change*
- [ ] New feature - *non-breaking change*
- [x] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

**Comment by mary-cleaton** [comment]
In draft - waiting til PR #63 is merged as needs to be rebased against that before being ready for QA.

File: configs_template.ini
Comment (danasebo) [inline configs_template.ini]: why is this file in this PR?

File: configs_template.ini [hunk 1]
```diff
@@ -0,0 +1,28 @@
+[run_spec]
+#Session size can be s, m, l or xl
+spark_session_size =
+
+[filepaths]
+# df1_path, df2_path and df_candidates_path must be to parquet files
```
Comment (danasebo) [inline configs_template.ini]: "[...] must be to parquet [...]" it sounds like a word is missing on this comment line

File: configs_template.ini [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# checkpoint_path must be to a unique location as it will be deleted at the end of the run
```
Comment (danasebo) [inline configs_template.ini]: "[...] must be to a unique location [...]" -> "must be a unique location"?

File: scalelink/main.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+"""
+Run script for Scalelink.
+"""
+
+import boto3
+import raz_client
+from rdsa_utils.cdp.helpers.s3_utils import delete_folder
```
Comment (danasebo) [inline scalelink/main.py:7]: `rdsa_utils` is not included in required packages in `setup.cfg` (and the script does not work unless `rdsa_utils` is installed) - is that intentional?

**Review by danasebo — CHANGES_REQUESTED** [review]
overall, all changes look good to me, great work! I tested both the unit tests and the main file, please look at the individual comments for minor requests. once addressed, the PR can be merged.

just to add that I tested that the `main.py` file runs without any issues, as well as the updated unit tests, and I also checked (using grep) that no other unnoticed instances of Python 3.8 / 3.9 are present in the repo.

File: configs_template.ini (hunk 1 above)
Comment (mary-cleaton) [reply configs_template.ini]: Corrected.

File: configs_template.ini (hunk 2 above)
Comment (mary-cleaton) [reply configs_template.ini]: Fixed, plus rephrased slightly to make it clearer.

File: scalelink/main.py (hunk 3 above)
Comment (mary-cleaton) [reply scalelink/main.py:7]: Drat. Not intentional!

I think there's actually several packages missing from the `setup.cfg`. Fixed now. Note, I've commented out PySpark as although it's needed, if you pip install it on our systems it'll break things. We'll need to undo that before publishing - I've put a note in our work record.

File: setup.cfg [hunk 4]
```diff
@@ -9,18 +9,22 @@ long_description_content_type = text/markdown
...
     Operating System :: OS Independent
 
 [options]
 packages = find:
-python_requires = >=3.8, <3.11
+python_requires = >=3.10, <3.12
 include_package_data = True
 install_requires =
-    pandas
+    boto3 >= 1.34.149
```
Comment (danasebo) [inline setup.cfg]: silly detail, but could you please update `>= 1.34.149` -> `>=1.34.149` (i.e. removing space after `>=` for consistency? (also for `pandas` & `raz_client`)

File: configs_template.ini [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+output_path =
+
+[variables]
+df1_id =
+df2_id =
+linkage_vars =
+df1_suffix =
+df2_suffix =
+
+[cutpoints]
```
Comment (danasebo) [inline configs_template.ini:23]: I think it is not clear that the individual cutpoints below need to be defined (i.e. if they are not to be used, they should be `None` rather than kept blank / undefined). I suggest to add this clarification as a comment to the template, if you agree

File: setup.cfg (hunk 4 above)
Comment (mary-cleaton) [reply setup.cfg]: Done.

File: configs_template.ini (hunk 5 above)
Comment (mary-cleaton) [reply configs_template.ini:23]: Done.

**Review by danasebo — APPROVED** [review]
All looks good to me now, thanks!
